runtime/io.c

```diff
@@ -252,7 +269,7 @@ CAMLexport int caml_flush_partial(struct channel *channel)
         /* This is a permanent failure: retrying the flush later will not
            make it go away.  Just discard the buffered data, so that
            the finalizer can reclaim the channel. */
-        channel->curr = channel->buff;
+        close_channel_buffer(channel);
```

Comment:
Perhaps worth clarifying that this doesn't rely on the OS returning `EBADF`,(which it may not, if the file descriptor got reused), but on OCaml's internal closed channel detection in 'struct channel' based on setting fd to -1.
Although maybe I'm wrong but 'close_channel_buffer' (as opposed to 'caml_ml_close_channel') won't actually set the fd to -1, and although I see that the comment moved in this PR says that a Sys error would be raised instead, I'm not sure if it'd always be. Should there be another flag "this channel is in permanent failed mode, the only thing you can do with it is close it?"